File: src/go/problems/flatten_test.go
// https://leetcode.com/problems/flatten-binary-tree-to-linked-list

package problems

import (
	"testing"
	"github.com/dobermanch/leetcode/core"
)

type Flatten struct{}

func TestFlatten(t *testing.T) {
	gen := core.TestSuite[Flatten]{}
	gen.Add(func(tc *core.TestCase) {
		root := TreeNode{
			Val: 1,
			Left: &TreeNode{ 
				Val: 2,
				Left: &TreeNode{ Val: 3 },
				Right: &TreeNode{ Val: 4 }
			},
			Right: &TreeNode{ 
				Val: 5,
				Right: &TreeNode{ Val: 6 }
			}
		}

		tc.Param(&root).Result(&root)
	}).Run(t)
}

func (Flatten) Solution1(root *TreeNode) *TreeNode {
	node := root
    for (node != nil) {
        if (node.Left != nil) {
            current := node.Left
            for (current.Right != nil) {
                current = current.Right                
            }

            current.Right = node.Right
            node.Right = node.Left
            node.Left = nil
        }

        node = node.Right
    }

	return root
}

func (Flatten) Solution2(root *TreeNode) *TreeNode {
	stack := []*TreeNode{}

	node := root
	var prev *TreeNode
	for (node != nil || len(stack) > 0) {
		if (node != nil) {
			stack = append(stack, node)
			prev = node
			node = node.Left
		} else {
			pop := stack[len(stack) - 1]
			stack = stack[:len(stack) - 1]
			node = pop.Right

			if (prev != nil) {
				prev.Right = node
				if (pop.Left != nil) {
					pop.Right = pop.Left
					pop.Left = nil
				}
			}
		}
	}

	return root
}

File: src/python/problems/flatten_Test.py
# https://leetcode.com/problems/flatten-binary-tree-to-linked-list

from typing import Optional
from core.problem_base import *
from models.tree_node import TreeNode

class CopyRandomList(ProblemBase):
    def Solution(self, root: Optional[TreeNode]) -> TreeNode:
        node = root
        while node:
            if node.left:
                current = node.left
                while current.right:
                    current = current.right
                
                current.right = node.right
                node.right = node.left
                node.left = None
            
            node = node.right

        return root


    def Solution1(self, root: Optional[TreeNode]) -> TreeNode:
        stack = []
        node = root
        prev: TreeNode = None
        while node or len(stack) > 0:
            if node:
                stack.append(node)
                prev = node
                node = node.left
            else:
                pop = stack.pop()
                node = pop.right

                if prev:
                    prev.right = node
                    if pop.left:
                        pop.right = pop.left
                        pop.left = None

        return root

if __name__ == '__main__':
    root = TreeNode(1, TreeNode(2, TreeNode(3), TreeNode(4)), TreeNode(5, None, TreeNode(6)))

    TestGen(CopyRandomList) \
        .Add(lambda tc: tc.Param(root).Result(root)) \
        .Run()
